---
title: "Genes best correlating with the selected gene"
output:
  html_document:
    toc: true
    # toc_float: true
    theme: united
    # theme: cerulean
    # number_sections: true
date: "`r Sys.Date()`"
author: "Mikhail Dozmorov"
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
# Set up the environment
library(knitr)
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=F, tidy=T, fig.keep='high', echo=F, dpi=100, warnings=F, message=F, comment=NA, warning=F, results='as.is', fig.width = 10, fig.height = 6) #out.width=700, 
library(pander)
panderOptions('table.split.table', Inf)
set.seed(1)
library(dplyr)
options(stringsAsFactors = FALSE)
```

```{r libraries}
library(yarn)
library(dplyr)
library(enrichR)
library(pathview)
```

```{r settings}
data_dir <- "/Users/mdozmorov/Documents/Data/GenomeRunner/GTEx/" # Mac
# data_dir <- "F:/Data/GenomeRunner/GTEx/" # Windows
# data_dir <- "data/" # Windows temporary

fileNameIn_gtex <- paste0(data_dir, "gtex.rds")
tissue <- "Lung" #  "Adipose Tissue", "Adrenal Gland", "Blood Vessel", "Bladder", "Brain", "Breast", "Blood", "Skin", "Cervix Uteri", "Colon", "Esophagus", "Fallopian Tube", "Heart", "Kidney", "Liver", "Lung", "Salivary Gland", "Muscle", "Nerve", "Ovary", "Pancreas", "Pituitary", "Prostate", "Small Intestine", "Spleen", "Stomach", "Testis", "Thyroid", "Uterus", "Vagina"
fileNameOut_gtex <- paste0(sub(".rds", "", fileNameIn_gtex), "_", tissue, ".rda") # Where to save subsetted and normalized object
proportion_cutoff <- 0.33 # maximum proportion of missing values, if more - exclude gene
corr_type <- "pearson" # Type of correlations, may be "spearman"
corr_cutoff <- 0.2 # Cutoff for selecting best correlated genes
fdr.cutoff <- 0.1 # Cutoff for functional enrichment analysis
extended_analysis <- FALSE # Run extended (=use all EnrichR databases) analysis?
lower_expr_percentile <- 0.1 # Lower expression percentile for the target gene, below which a sample is excluded
top_pathways_to_plot  <- 10 # How many top KEGG pathways to plot?
```

```{r}
library(annotables)
# Remove non-canonical chromosome names
grch38 <- grch38[ !(grepl("_", grch38$chr) | grepl("GL", grch38$chr)), ]
grch38 <- grch38[, c("ensgene", "symbol", "description")]
grch38 <- grch38[ !duplicated(grch38), ]

# A wrapper function to perform all functional enrichment analyses.
# Helper function to save non-empty results
save_res <- function(res, fileName = fileName, wb = wb, sheetName = "KEGG") {
  if (nrow(res) > 0) {
    openxlsx::addWorksheet(wb = wb, sheetName = sheetName)
    openxlsx::writeData(wb, res, sheet = sheetName)
    openxlsx::saveWorkbook(wb, fileName, overwrite = TRUE)
  }
}

# A wrapper to save the results
save_enrichr <- function(up.genes = up.genes, dn.genes = dn.genes, databases = "KEGG_2016", fdr.cutoff = fdr.cutoff, fileNameOut = fileNameOut, wb = wb) {
  print(paste("Running", databases, "analysis", sep = " "))
  res.kegg <- enrichFullGeneList(up.genes, dn.genes, databases = databases, fdr.cutoff = fdr.cutoff)
  res.kegg$pval <- formatC(res.kegg$pval, digits = 3, format = "e")
  res.kegg$qval <- formatC(res.kegg$qval, digits = 3, format = "e")
  if (nchar(databases) > 30) databases <- paste0(substr(databases, 1, 20), "_", substr(databases, nchar(databases) - 8, nchar(databases))) # If a database is longer that 30 characters, keep first 20 and last 10 characters
  save_res(res.kegg, fileNameOut, wb = wb, sheetName = databases)
  # Pause for a few seconds
  pause_sec <- round(runif(1, min = 1, max = 10))
  Sys.sleep(pause_sec)
  return(res.kegg)
}
```

```{r load}
if (!file.exists(fileNameOut_gtex)) { # If not preprocessed file
  # load the original file
  obj <- readRDS(file = fileNameIn_gtex)
  # subset
  ind_tissue <- which(pData(obj)$SMTS == tissue)
  obj <- obj[, ind_tissue ]
  # # filter
  # ff <- genefilter::pOverA(p = proportion_cutoff, A = 0, na.rm = TRUE) # Should be more than 90% of non-zero values
  # obj <- obj[apply(exprs(obj), 1, ff), ] 
  # normalize log2-transformed data
  exprs <- limma::normalizeQuantiles(log2(exprs(obj) + 1))
  exprs(obj) <- exprs
  # Save
  save(obj, file = fileNameOut_gtex) # Select cancers
} else {
  load(file = fileNameOut_gtex)
}
```

```{r select_genes}
# Select by gene name
# selected_genes <- c("LOC151121")
# sum(fData(obj)$hgnc_symbol %in% selected_genes)
# Or, select by Ensembl ID
# selected_genes <- c("ENSG00000250658") # LOC339975
# selected_genes <- c("ENSG00000204460") # LOC151121
# selected_genes <- c("ENSG00000246090") # LOC100507053
# selected_genes <- c("ENSG00000115884") # selected_names <- c("SDC1")
selected_genes <- c("ENSG00000090932") # DLL3
# selected_genes <- c("ENSG00000175279") # APITD1-CORT

sum(fData(obj)$ensembl_gene_id %in% selected_genes)
fData(obj)[ fData(obj)$ensembl_gene_id %in% selected_genes, ]
# Goal - to get index of the selected gene
ind_gene <- which(fData(obj)$ensembl_gene_id %in% selected_genes)
```

```{r prepareFile}
# File name to save all results
fileNameOut <- paste0("results/GTEx_", tissue, "_", selected_genes, ".xlsx")
unlink(fileNameOut)
wb <- openxlsx::createWorkbook(fileNameOut) # openxlsx::loadWorkbook(fileName)
```

# Correlation analysis

We have `r ncol(exprs(obj))` samples to run correlation on

```{r correlations}
# File name to save correlation object
fileNameOut_corr <- paste0("data/corr_", selected_genes, ".rda")

if (!file.exists(fileNameOut_corr)) { # If correlations are not precomputed
  max_corrs <- nrow(obj) # How many rows to process
  all_corrs <- vector(mode = "numeric", length = max_corrs)
  all_pvals <- vector(mode = "numeric", length = max_corrs)
  target_gene <- as.numeric(exprs(obj)[ ind_gene, ]) # The target gene to run correlation on
  idx <- which(target_gene > quantile(target_gene, probs = lower_expr_percentile)) # Which are non-zero?
  target_gene <- target_gene[ idx ]
  exprs <- exprs(obj)[, idx] # Extract expression matrix once, subsetted
  for (i in 1:max_corrs) {
    cors         <- Hmisc::rcorr(target_gene, as.numeric(exprs[ i, ]), type = corr_type)
    all_corrs[i] <- cors[[1]][1, 2]
    all_pvals[i] <- cors[[3]][1, 2]
  }
  correlations <- data.frame(fData(obj)[1:max_corrs, ], corr = all_corrs, pval = all_pvals) # Assemble correlations
  correlations <- left_join(correlations, grch38, by = c("ensembl_gene_id" = "ensgene"))
  save(correlations, file = fileNameOut_corr)
} else {
  load(file = fileNameOut_corr)
}
save_res(correlations[ order(correlations$corr, decreasing = TRUE), ], fileNameOut, wb = wb, sheetName = "Correlations")
```

If we remove samples having expression of the target gene `r selected_genes` below `r lower_expr_percentile` percentile, we are left with `r ncol(exprs)` samples

## Positively correlated genes 


Genes positively correlated with the selected gene `r selected_genes` across all `r tissue` tissues. Correlation method: `r corr_type`, correlation coefficient cutoff: >`r corr_cutoff`. Legend:

- `symbol`, `description` - gene symbols/description
- `cor`, `pval` - Pearson correlation coefficient, and p-value of correlation significance

```{r}
correlations_pos <- correlations[ correlations$corr > corr_cutoff, ]
correlations_pos$corr <- signif(correlations_pos$corr)
correlations_pos$pval <- signif(correlations_pos$pval)
DT::datatable(correlations_pos[ order(correlations_pos$corr, decreasing = TRUE), ])
up.genes <- unique(correlations_pos$hgnc_symbol[ !(is.na(correlations_pos$hgnc_symbol) | correlations_pos$hgnc_symbol == "") ])
```

We have `r nrow(correlations_pos)` genes positively correlated with `r selected_genes`.

## Negatively correlated genes

```{r}
correlations_neg <- correlations[ correlations$corr < -corr_cutoff, ]
correlations_neg$corr <- signif(correlations_neg$corr)
correlations_neg$pval <- signif(correlations_neg$pval)
DT::datatable(correlations_neg[ order(correlations_neg$corr, decreasing = FALSE), ])
dn.genes <- unique(correlations_neg$hgnc_symbol[ !(is.na(correlations_neg$hgnc_symbol) | correlations_neg$hgnc_symbol == "") ])
```

We have `r nrow(correlations_neg)` genes negatively correlated with `r selected_genes`.


# Functional enrichment analysis

- Each table has enrichment results separately for up/downregulated genes. The "direction" column indicate which pathways are enriched in "UP"- or "DN"-regulated genes.

- Use the "Search" box for each table, to filter the results for "UP" or "DN" only. Search is global within the table, case insensitive.

- FDR cutoff - `r fdr.cutoff`.

## KEGG

```{r}
res.kegg <- save_enrichr(up.genes = up.genes, dn.genes = dn.genes, databases = "KEGG_2016", fdr.cutoff = fdr.cutoff, fileNameOut = fileNameOut, wb = wb)
DT::datatable(res.kegg)
```

```{r eval=FALSE}
# Save named correlation vector
# For upregulated
up.degs.genes <- aggregate(correlations_neg$corr, by = list(gene = correlations_neg$hgnc_symbol), FUN = mean) # Aggregate potential duplicated names
up.degs.genes <- up.degs.genes[ !(up.degs.genes$gene == "" | is.na(up.degs.genes)), ] # Remove empty gene names
up.degs.names <- up.degs.genes$gene
up.degs.genes <- up.degs.genes$x
names(up.degs.genes) <- up.degs.names
# and downregulated genes
dn.degs.genes <- aggregate(correlations_neg$corr, by = list(gene = correlations_neg$hgnc_symbol), FUN = mean) # Aggregate potential duplicated names
dn.degs.genes <- dn.degs.genes[ !(dn.degs.genes$gene == "" | is.na(dn.degs.genes)), ] # Remove empty gene names
dn.degs.names <- dn.degs.genes$gene
dn.degs.genes <- dn.degs.genes$x
names(dn.degs.genes) <- dn.degs.names
# Combine into one
degs.genes <- c(up.degs.genes, dn.degs.genes)

# Get pathways
# For upregulated
up.res.kegg <- res.kegg$category[ res.kegg$direction == "UP" ]
if (length(up.res.kegg) > 0) {
  up.res.kegg <- up.res.kegg[ grepl("sapiens", up.res.kegg) ]
  up.res.kegg.names <- sapply(up.res.kegg, function(x) { strsplit(x, "_Homo sapiens_hsa")[[1]][1] })
  up.res.kegg.ids   <- sapply(up.res.kegg, function(x) { strsplit(x, "_hsa")[[1]][2] })
}
# and downregulated genes
dn.res.kegg <- res.kegg$category[ res.kegg$direction == "DN" ]
if (length(dn.res.kegg) > 0) {
  dn.res.kegg <- dn.res.kegg[ grepl("sapiens", dn.res.kegg) ]
  dn.res.kegg.names <- sapply(dn.res.kegg, function(x) { strsplit(x, "_Homo sapiens_hsa")[[1]][1] })
  dn.res.kegg.ids   <- sapply(dn.res.kegg, function(x) { strsplit(x, "_hsa")[[1]][2] })
}

# Get pathways
for (i in 1:top_pathways_to_plot) {
  if (!is.na(up.res.kegg.ids[i])) {
    pathview(gene.data = degs.genes, pathway.id = up.res.kegg.ids[i], species = "hsa", gene.idtype = "SYMBOL", gene.annotpkg = "org.Hs.eg.db", out.suffix = paste0(up.res.kegg.names[i], "_UP"))
  }
  if (!is.na(dn.res.kegg.ids[i])) {
    pathview(gene.data = degs.genes, pathway.id = dn.res.kegg.ids[i], species = "hsa", gene.idtype = "SYMBOL", gene.annotpkg = "org.Hs.eg.db", out.suffix = paste0(dn.res.kegg.names[i], "_DN"))
  }
}
```

