---
title: "Genes best correlating with the selected gene"
output:
  html_document:
    toc: true
    # toc_float: true
    theme: united
    # theme: cerulean
    # number_sections: true
date: "`r Sys.Date()`"
author: "Mikhail Dozmorov"
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
# Set up the environment
library(knitr)
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=F, tidy=T, fig.keep='high', echo=F, dpi=100, warnings=F, message=F, comment=NA, warning=F, results='as.is', fig.width = 10, fig.height = 6) #out.width=700, 
library(pander)
panderOptions('table.split.table', Inf)
set.seed(1)
library(dplyr)
options(stringsAsFactors = FALSE)
```

```{r libraries}
library(yarn)
library(dplyr)
library(enrichR)
library(pathview)
```

```{r settings}
data_dir <- "/Users/mdozmorov/Documents/Data/GenomeRunner/GTEx/" # Mac
# data_dir <- "F:/Data/GenomeRunner/GTEx/" # Windows
# data_dir <- "data/" # Windows temporary

fileNameIn_gtex <- paste0(data_dir, "gtex.rds")
tissue <- c("Adipose Tissue", "Adrenal Gland", "Blood Vessel", "Bladder", "Brain", "Breast", "Blood", "Skin", "Cervix Uteri", "Colon", "Esophagus", "Fallopian Tube", "Heart", "Kidney", "Liver", "Lung", "Salivary Gland", "Muscle", "Nerve", "Ovary", "Pancreas", "Pituitary", "Prostate", "Small Intestine", "Spleen", "Stomach", "Testis", "Thyroid", "Uterus", "Vagina")

proportion_cutoff <- 0.33 # maximum proportion of missing values, if more - exclude gene
corr_type <- "pearson" # Type of correlations, may be "spearman"
corr_cutoff <- 0.2 # Cutoff for selecting best correlated genes
fdr.cutoff <- 0.1 # Cutoff for functional enrichment analysis
extended_analysis <- FALSE # Run extended (=use all EnrichR databases) analysis?
lower_expr_percentile <- 0.1 # Lower expression percentile for the target gene, below which a sample is excluded
top_pathways_to_plot  <- 10 # How many top KEGG pathways to plot?
```

```{r}
library(annotables)
# Remove non-canonical chromosome names
grch38 <- grch38[ !(grepl("_", grch38$chr) | grepl("GL", grch38$chr)), ]
grch38 <- grch38[, c("ensgene", "symbol", "description")]
grch38 <- grch38[ !duplicated(grch38), ]

# A wrapper function to perform all functional enrichment analyses.
# Helper function to save non-empty results
save_res <- function(res, fileName = fileName, wb = wb, sheetName = "KEGG") {
  if (nrow(res) > 0) {
    openxlsx::addWorksheet(wb = wb, sheetName = sheetName)
    openxlsx::writeData(wb, res, sheet = sheetName)
    openxlsx::saveWorkbook(wb, fileName, overwrite = TRUE)
  }
}

# A wrapper to save the results
save_enrichr <- function(up.genes = up.genes, dn.genes = dn.genes, databases = "KEGG_2016", fdr.cutoff = fdr.cutoff, fileNameOut = fileNameOut, wb = wb, tissue = tissue) {
  print(paste("Running", databases, "analysis", sep = " "))
  res.kegg <- enrichFullGeneList(up.genes, dn.genes, databases = databases, fdr.cutoff = fdr.cutoff)
  res.kegg$pval <- formatC(res.kegg$pval, digits = 3, format = "e")
  res.kegg$qval <- formatC(res.kegg$qval, digits = 3, format = "e")
  # Format sheetName
  sheetName <- paste0(tissue, "_", databases)
  if (nchar(sheetName) > 30) sheetName <- paste0(substr(sheetName, 1, 20), "_", substr(sheetName, nchar(sheetName) - 8, nchar(sheetName))) # If a sheetName is longer that 30 characters, keep first 20 and last 10 characters
  save_res(res.kegg, fileNameOut, wb = wb, sheetName = sheetName)
  # Pause for a few seconds
  pause_sec <- round(runif(1, min = 1, max = 10))
  Sys.sleep(pause_sec)
  return(res.kegg)
}
```

```{r load}
# load the original file
obj <- readRDS(file = fileNameIn_gtex)
for (t in tissue) {
  fileNameOut_gtex <- paste0(sub(".rds", "", fileNameIn_gtex), "_", t, ".rda") # Where to save subsetted and normalized object
  if (!file.exists(fileNameOut_gtex)) { # If not preprocessed file
    print(paste0("Processing ", t))
    # subset
    ind_tissue <- which(pData(obj)$SMTS == t)
    obj_tissue <- obj[, ind_tissue ]
    # filter
    ff <- genefilter::pOverA(p = proportion_cutoff, A = 0, na.rm = TRUE) # Should be more than 90% of non-zero values
    obj_tissue <- obj_tissue[apply(exprs(obj_tissue), 1, ff), ] 
    # normalize log2-transformed data
    exprs <- limma::normalizeQuantiles(log2(exprs(obj_tissue) + 1))
    exprs(obj_tissue) <- exprs
    # Save
    save(obj_tissue, file = fileNameOut_gtex) # Select cancers
  }
}
```

```{r select_genes}
# Select by gene name
# selected_genes <- c("LOC151121")
# sum(fData(obj)$hgnc_symbol %in% selected_genes)
# Or, select by Ensembl ID
# selected_genes <- c("ENSG00000250658") # LOC339975
# selected_genes <- c("ENSG00000204460") # LOC151121
# selected_genes <- c("ENSG00000246090") # LOC100507053
selected_genes <- c("ENSG00000136485") # DCAF7

sum(fData(obj)$ensembl_gene_id %in% selected_genes)
fData(obj)[ fData(obj)$ensembl_gene_id %in% selected_genes, ]
```

# Correlation analysis

We have `r ncol(exprs(obj))` samples to run correlation on

```{r correlations}
for (t in tissue) {
  # File name to save correlation object
  fileNameOut_corr <- paste0("data/corr_", selected_genes, "_", t, ".rda")
  if (!file.exists(fileNameOut_corr)) { # If correlations are not precomputed
    # Load Tissue-specific expression
    fileNameOut_gtex <- paste0(sub(".rds", "", fileNameIn_gtex), "_", t, ".rda") # Where to save subsetted and normalized object
    load(file = fileNameOut_gtex)
    # Goal - to get index of the selected gene
    ind_gene <- which(fData(obj_tissue)$ensembl_gene_id %in% selected_genes)
    if (length(ind_gene) == 1) {
      print(paste0("Processing ", t))
      max_corrs <- nrow(obj_tissue) # How many rows to process
      all_corrs <- vector(mode = "numeric", length = max_corrs)
      all_pvals <- vector(mode = "numeric", length = max_corrs)
      target_gene <- as.numeric(exprs(obj_tissue)[ ind_gene, ]) # The target gene to run correlation on
      idx <- which(target_gene > quantile(target_gene, probs = lower_expr_percentile)) # Which are non-zero?
      target_gene <- target_gene[ idx ]
      exprs <- exprs(obj_tissue)[, idx] # Extract expression matrix once, subsetted
      for (i in 1:max_corrs) {
        cors         <- Hmisc::rcorr(target_gene, as.numeric(exprs[ i, ]), type = corr_type)
        all_corrs[i] <- cors[[1]][1, 2]
        all_pvals[i] <- cors[[3]][1, 2]
      }
      correlations <- data.frame(fData(obj_tissue)[1:max_corrs, ], corr = all_corrs, pval = all_pvals) # Assemble correlations
      correlations <- left_join(correlations, grch38, by = c("ensembl_gene_id" = "ensgene"))
      save(correlations, file = fileNameOut_corr)      
    } else {
      print("Selected gene either is not present, or matches multiple IDs")
      fData(obj_tissue)[ fData(obj_tissue)$ensembl_gene_id %in% selected_genes, ]
    }
  } 
}
```

Genes correlated with the selected gene `r selected_genes` in each tissue. Correlation method: `r corr_type`, correlation coefficient cutoff: >`r corr_cutoff`. Legend:

- `symbol`, `description` - gene symbols/description
- `cor`, `pval` - Pearson correlation coefficient, and p-value of correlation significance

```{r}
# File name to save all results
fileNameOut <- paste0("results/GTEx_", "_", selected_genes, ".xlsx")
unlink(fileNameOut)
wb <- openxlsx::createWorkbook(fileNameOut) # openxlsx::loadWorkbook(fileName)

for (t in tissue) {
  # File name to save correlation object
  fileNameOut_corr <- paste0("data/corr_", selected_genes, "_", t, ".rda")
  load(file = fileNameOut_corr)
  save_res(correlations[ order(correlations$corr, decreasing = TRUE), ], fileNameOut, wb = wb, sheetName = paste0("CORR_", t))
  # Get positive and negative correlated genes
  correlations_pos <- correlations[ correlations$corr > corr_cutoff, ]
  up.genes <- unique(correlations_pos$hgnc_symbol[ !(is.na(correlations_pos$hgnc_symbol) | correlations_pos$hgnc_symbol == "") ])
  correlations_neg <- correlations[ correlations$corr < -corr_cutoff, ]
  dn.genes <- unique(correlations_neg$hgnc_symbol[ !(is.na(correlations_neg$hgnc_symbol) | correlations_neg$hgnc_symbol == "") ])
  print(paste0("Processing ", length(up.genes), " positively and ", length(dn.genes), " negatively correlated genes, in ", t, " tissue"))
  # KEGG
  res.kegg <- save_enrichr(up.genes = up.genes, dn.genes = dn.genes, databases = "KEGG_2016", fdr.cutoff = fdr.cutoff, fileNameOut = fileNameOut, wb = wb, tissue = t)
}
```
